clang/docs/SYCLFERFCLoweringKernel.md

```diff
@@ -0,0 +1,88 @@
+# SYCL Kernel
+
+A SYCL construct such as 'single_task' takes a function object as one of its arguments.   The contents of this function object are executed on the device.   To enable this, the function object is converted into the format of an OpenCL kernel.
+
+Consider the following code snippet:
+
+```
+  sycl::accessor<char, 1, sycl::access::mode::read> acc1;
+  int i = 13;
+  sycl::sampler smplr;
+
+  struct S {
+    char c;
+    int i;
+  } test_s;
+
+  test_s.c = 14;
+
+  kernel_single_task<class first_kernel>([=]() {
+    if (i == 13 && test_s.c == 14) {
+
+      acc1.use();
+      smplr.use();
+    }
+  });
+```
+
+Here the lambda within the 'kernel_single_task' construct needs to be executed on the device.  To do this, a function object representing the kernel is generated.   The function call operator of this object has the contents of the kernel invocation.  The function object generated looks like:
+
+```
+struct FuncObj {
+  int i;
+  struct S test_s;
+  sycl::accessor acc1;
+  sycl::sampler smplr;
+
+  () {  // Function call operator
+    if (i == 13 && test_s.c == 14) {
+      acc1.use();
+      smplr.use();
+    }
+  }
+};
+```
+
+The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.
```

Comment:
Might be good to expand a bit on the process, from experience many get confused by it.

```suggestion
The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.
It does so by walking the function object data member and producing a parameter for each of them.
Inside the OpenCL kernel, the function object is rebuilt and then called.
Some special types like accessor and sampler are treated a bit differently (see below). 
```